File: cash_flow_app/cash_flow_management/report/operational_balance_sheet/operational_balance_sheet.js
// Copyright (c) 2025, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["Operational Balance Sheet"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("Sana dan"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start(),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("Sana gacha"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_end(),
			"reqd": 1
		},
		{
			"fieldname": "periodicity",
			"label": __("Davr"),
			"fieldtype": "Select",
			"options": ["Daily", "Monthly", "Quarterly", "Yearly"],
			"default": "Monthly",
			"reqd": 1
		}
	],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Main sections (AKTIVLAR, JAMI KREDITORKA, BALANS)
		if (data && data.indent === 0) {
			if (column.fieldname === "account") {
				value = `<span style="font-weight: bold; font-size: 15px; color: #2c3e50; text-transform: uppercase;">${value}</span>`;
			} else {
				value = `<span style="font-weight: bold; font-size: 14px;">${value}</span>`;
			}
		}

		// Level 1 bold (Pul, Debitorka, Kreditorka, Ustav, Sof Foyda)
		if (data && data.indent === 1 && data.is_group) {
			value = `<span style="font-weight: bold;">${value}</span>`;
		}

		// Level 2 bold (Mijozlar, Yetkazib beruvchilar)
		if (data && data.indent === 2 && data.is_group) {
			value = `<span style="font-weight: bold;">${value}</span>`;
		}

		// Level 3 bold (Customer Groups, Supplier Groups)
		if (data && data.indent === 3 && data.is_group) {
			value = `<span style="font-weight: bold; color: #555;">${value}</span>`;
		}

		// BALANS row highlighting
		if (data && data.account && data.account === "BALANS") {
			if (column.fieldname !== "account") {
				const numValue = parseFloat(data[column.fieldname]) || 0;
				const color = Math.abs(numValue) < 0.01 ? "#27ae60" : "#e74c3c";
				value = `<span style="font-weight: bold; color: ${color}; font-size: 14px;">${value}</span>`;
			}
		}

		// Format negative numbers with minus sign
		if (column.fieldname !== "account" && data) {
			const numValue = parseFloat(data[column.fieldname]);
			if (!isNaN(numValue) && numValue < 0) {
				value = value.replace(/[\d,.-]+/, function(match) {
					const absValue = Math.abs(parseFloat(match.replace(/,/g, '')));
					return `-${absValue.toLocaleString('en-US', {
						minimumFractionDigits: 2,
						maximumFractionDigits: 2
					})}`;
				});
				value = `<span style="color: #e74c3c;">${value}</span>`;
			}
		}

		return value;
	},

	"tree": true,
	"name_field": "account",
	"parent_field": "",
	"initial_depth": 2,

	onload: function(report) {
		// Refresh button
		report.page.add_inner_button(__("Yangilash"), function() {
			report.refresh();
		});

		// Export to Excel
		report.page.add_inner_button(__("Excel'ga eksport"), function() {
			frappe.utils.to_excel(report.data, report.report_name);
		});

		// Help button
		report.page.set_secondary_action(__("Yordam"), function() {
			frappe.msgprint({
				title: __("Operatsion Balans Hisoboti - Yordam"),
				message: `
					<h4 style="color: #2c3e50; margin-top: 15px;">📊 Hisobot Strukturasi</h4>
					<ul style="line-height: 1.8;">
						<li><b>AKTIVLAR:</b>
							<ul>
								<li>Pul - Kassa va bank hisoblaridagi mablag'lar</li>
								<li>Debitorka:
									<ul>
										<li>Mijozlar - Nasiya sotilgan mahsulotlar</li>
										<li>Yetkazib beruvchilar - Berilgan avanslar</li>
									</ul>
								</li>
							</ul>
						</li>
						<li style="margin-top: 10px;"><b>JAMI KREDITORKA:</b>
							<ul>
								<li>Kreditorka:
									<ul>
										<li>Mijozlar - Olingan avanslar</li>
										<li>Yetkazib beruvchilar - Yetkazib beruvchilarga qarzlar</li>
									</ul>
								</li>
								<li>Ustav Kapitali - Aksiyadorlardan olingan mablag'lar</li>
								<li>Sof Foyda - Foiz daromadlari minus xarajatlar</li>
							</ul>
						</li>
						<li style="margin-top: 10px;"><b>BALANS:</b>
							<span style="color: #27ae60;">Aktivlar - Jami Kreditorka = 0</span>
							(agar to'g'ri hisoblanmagan bo'lsa, qizil rangda ko'rinadi)
						</li>
					</ul>

					<h4 style="color: #2c3e50; margin-top: 20px;">📅 Davr Turlari</h4>
					<ul style="line-height: 1.8;">
						<li><b>Daily (Kunlik):</b> Har bir kun uchun alohida ustun</li>
						<li><b>Monthly (Oylik):</b> Har bir oy uchun alohida ustun</li>
						<li><b>Quarterly (Choraklik):</b> Har bir chorak uchun alohida ustun</li>
						<li><b>Yearly (Yillik):</b> Har bir yil uchun alohida ustun</li>
					</ul>

					<h4 style="color: #2c3e50; margin-top: 20px;">💾 Ma'lumotlar Manbalari</h4>
					<ul style="line-height: 1.8;">
						<li><b>Pul:</b> Payment Entry (Receive/Pay/Internal Transfer)</li>
						<li><b>Debitorka (Mijozlar):</b> Sales Order + Payment Entry (Pay)</li>
						<li><b>Debitorka (Yetkazib beruvchilar):</b> Payment Entry (Supplier, Pay)</li>
						<li><b>Kreditorka (Mijozlar):</b> Payment Entry (Customer, Receive)</li>
						<li><b>Kreditorka (Yetkazib beruvchilar):</b> Payment Entry (Supplier, Receive)</li>
						<li><b>Ustav Kapitali:</b> Payment Entry (Shareholder) - Receive minus Pay</li>
						<li><b>Sof Foyda:</b> Installment Application (Foiz daromadlari) - Counterparty Category (Xarajatlar)</li>
					</ul>

					<h4 style="color: #2c3e50; margin-top: 20px;">🎨 Ranglar</h4>
					<ul style="line-height: 1.8;">
						<li><span style="color: #27ae60;">● Yashil</span> - BALANS 0 ga teng (to'g'ri)</li>
						<li><span style="color: #e74c3c;">● Qizil</span> - BALANS 0 ga teng emas (xatolik bor)</li>
						<li><span style="color: #e74c3c;">● Qizil raqamlar</span> - Manfiy qiymatlar</li>
					</ul>

					<p style="margin-top: 20px; padding: 10px; background: #f8f9fa; border-left: 3px solid #3498db;">
						<b>💡 Maslahat:</b> Qatorlarni ochish/yopish uchun chap tomondagi + / - tugmalarini bosing.
					</p>
				`,
				wide: true
			});
		});

		// Add success message after load
		frappe.show_alert({
			message: __('Hisobot muvaffaqiyatli yuklandi'),
			indicator: 'green'
		}, 3);
	}
};

File: cash_flow_app/cash_flow_management/report/operational_balance_sheet/operational_balance_sheet.py
# Copyright (c) 2025, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, add_days, get_last_day, get_first_day
from datetime import datetime
from collections import defaultdict


def execute(filters=None):
	"""Main execution function"""
	if not filters:
		filters = {}

	validate_filters(filters)
	columns = get_columns(filters)
	data_loader = OperationalBalanceData(filters)
	raw_data = data_loader.load_all_data()
	data = build_tree_structure(raw_data, filters)

	return columns, data


def validate_filters(filters):
	"""Validate and set default filters"""
	if not filters.get("from_date"):
		filters["from_date"] = get_first_day(getdate())
	if not filters.get("to_date"):
		filters["to_date"] = get_last_day(getdate())
	if not filters.get("periodicity"):
		filters["periodicity"] = "Monthly"

	filters["from_date"] = getdate(filters["from_date"])
	filters["to_date"] = getdate(filters["to_date"])

	if filters["from_date"] > filters["to_date"]:
		frappe.throw(_("From Date cannot be after To Date"))


def get_columns(filters):
	"""Generate dynamic columns based on periodicity"""
	columns = [
		{
			"label": _("Hisob"),
			"fieldname": "account",
			"fieldtype": "Data",
			"width": 350
		}
	]

	periods = get_periods(filters)
	for period in periods:
		columns.append({
			"label": period["label"],
			"fieldname": period["key"],
			"fieldtype": "Currency",
			"width": 150
		})

	return columns


def get_periods(filters):
	"""Generate period list based on periodicity"""
	periods = []
	periodicity = filters.get("periodicity", "Monthly")
	from_date = filters["from_date"]
	to_date = filters["to_date"]
	current_date = from_date

	if periodicity == "Daily":
		while current_date <= to_date:
			periods.append({
				"key": current_date.strftime("%Y-%m-%d"),
				"label": current_date.strftime("%d/%m/%Y"),
				"from_date": current_date,
				"to_date": current_date
			})
			current_date = add_days(current_date, 1)

	elif periodicity == "Monthly":
		while current_date <= to_date:
			month_end = get_last_day(current_date)
			if month_end > to_date:
				month_end = to_date
			periods.append({
				"key": current_date.strftime("%Y-%m"),
				"label": current_date.strftime("%b %Y"),
				"from_date": get_first_day(current_date),
				"to_date": month_end
			})
			next_month = add_days(month_end, 1)
			current_date = next_month

	elif periodicity == "Quarterly":
		quarter_start = get_quarter_start(from_date)
		while quarter_start <= to_date:
			quarter_end = get_quarter_end(quarter_start)
			if quarter_end > to_date:
				quarter_end = to_date
			quarter_num = ((quarter_start.month - 1) // 3) + 1
			periods.append({
				"key": f"{quarter_start.year}-Q{quarter_num}",
				"label": f"Q{quarter_num} {quarter_start.year}",
				"from_date": quarter_start,
				"to_date": quarter_end
			})
			quarter_start = add_days(quarter_end, 1)

	elif periodicity == "Yearly":
		year = from_date.year
		while year <= to_date.year:
			year_start = datetime(year, 1, 1).date()
			year_end = datetime(year, 12, 31).date()
			if year_start < from_date:
				year_start = from_date
			if year_end > to_date:
				year_end = to_date
			periods.append({
				"key": str(year),
				"label": str(year),
				"from_date": year_start,
				"to_date": year_end
			})
			year += 1

	return periods


def get_quarter_start(date):
	"""Get quarter start date"""
	quarter = ((date.month - 1) // 3) + 1
	month = (quarter - 1) * 3 + 1
	return datetime(date.year, month, 1).date()


def get_quarter_end(quarter_start):
	"""Get quarter end date"""
	month = quarter_start.month + 2
	year = quarter_start.year
	if month > 12:
		month = 12
	return get_last_day(datetime(year, month, 1).date())


class OperationalBalanceData:
    """Optimized data loader - GL Entry based"""

    def __init__(self, filters):
        self.filters = filters
        self.to_date = filters["to_date"]

    def load_all_data(self):
        """Hamma narsani yuklash: GL (raqamlar) + Categories (guruhlar)"""
        return {
            "gl_entries": self.load_gl_data(),
            # Quyidagi ikkita funksiya pastda aniqlangan bo'lishi SHART
            "counterparty_categories": self.load_counterparty_categories(),
            "shareholders": self.load_shareholders(),
            "cash_accounts": self.load_cash_accounts()
        }

    def load_gl_data(self):
        """
        Barcha moliyaviy haqiqat manbasi (GL Entry).
        """
        return frappe.db.sql("""
            SELECT
                party_type,
                party,
                account,
                SUM(debit) as total_debit,
                SUM(credit) as total_credit,
                (SUM(debit) - SUM(credit)) as balance
            FROM `tabGL Entry`
            WHERE posting_date <= %(to_date)s
              AND is_cancelled = 0
            GROUP BY party_type, party, account
        """, {"to_date": self.to_date}, as_dict=1)

    def load_counterparty_categories(self):
        """Sizda yetishmayotgan funksiya: Kustom kategoriyalarni yuklash"""
        return frappe.db.sql("""
            SELECT name, category_name, category_type, custom_expense_type
            FROM `tabCounterparty Category`
            WHERE is_active = 1
        """, as_dict=1)

    def load_shareholders(self):
        """Ta'sischilarni yuklash"""
        return frappe.db.sql("""
            SELECT name, title, custom_category
            FROM `tabShareholder`
        """, as_dict=1)

    def load_cash_accounts(self):
        """Kassa hisob raqamlarini yuklash (Pul bo'limi uchun)"""
        # Buni GL dan olish ham mumkin, lekin hozircha eski koddan qoldiramiz
        # sababi build_tree_structure da 'cash_accounts' ishlatilyapti
        return frappe.db.sql("""
            SELECT name as account_name
            FROM `tabAccount`
            WHERE account_type = 'Cash' AND is_group = 0
        """, as_dict=1)
def build_tree_structure(raw_data, filters):
	"""
	YANGILANGAN MANTIQ: GL Entry asosida Balans
	"""
	periods = get_periods(filters)
	data = []

	# ---------------------------------------------------------
	# YORDAMCHI FUNKSIYA: Qator yaratish
	# ---------------------------------------------------------
	def create_row(account, indent=0, is_group=False, **kwargs):
		row = {"account": account, "indent": indent, "is_group": is_group}
		# Hamma periodlar uchun 0.0 qo'yib chiqamiz
		for period in periods:
			row[period["key"]] = 0.0
		row.update(kwargs)
		return row

	# ---------------------------------------------------------
	# 1. MA'LUMOTLARNI TAYYORLASH
	# ---------------------------------------------------------
	gl_entries = raw_data.get("gl_entries", [])

	# Category Mapni to'g'rilab olamiz (List -> Dict)
	# Agar load_counterparty_categories list qaytarsa, uni dict qilamiz: {'PartyName': 'CategoryName'}
	categories_raw = raw_data.get("counterparty_categories", [])
	categories_map = {}
	if isinstance(categories_raw, list):
		for c in categories_raw:
			# Agar c dict bo'lsa
			if isinstance(c, dict) and "name" in c:
				categories_map[c["name"]] = c.get("category_name", "Boshqa")
	elif isinstance(categories_raw, dict):
		categories_map = categories_raw

	# Lug'atlarni tayyorlash
	supplier_payables = defaultdict(dict)  # Kreditorka (Qarzlar) -> Passiv
	supplier_advances = defaultdict(dict)  # Debitorka (Avanslar) -> Aktiv
	customer_receivables = defaultdict(dict)  # Debitorka (Qarzlar) -> Aktiv
	customer_advances = defaultdict(dict)  # Kreditorka (Avanslar) -> Passiv

	# ---------------------------------------------------------
	# 2. GL ENTRYLARNI GURUHLASH (SIZ BERGAN MANTIQ)
	# ---------------------------------------------------------
	for entry in gl_entries:
		party = entry.party
		balance = flt(entry.balance)

		if balance == 0:
			continue

		# --- YETKAZIB BERUVCHILAR ---
		if entry.party_type == "Supplier":
			# Guruhni aniqlash
			if party in categories_map:
				group = categories_map[party]
			else:
				group = frappe.get_cached_value("Supplier", party, "supplier_group") or "Boshqa"

			# GL Balans: (Debit - Credit)
			# Manfiy = Biz Qarzmiz (Kreditorka)
			# Musbat = Biz Haqdorimiz (Debitorka/Avans)
			if balance < 0:
				if party not in supplier_payables[group]:
					supplier_payables[group][party] = 0.0
				supplier_payables[group][party] += abs(balance)
			else:
				# "Nomalum Pul" shu yerga tushadi ($20.29)
				if party not in supplier_advances[group]:
					supplier_advances[group][party] = 0.0
				supplier_advances[group][party] += abs(balance)

		# --- MIJOZLAR ---
		elif entry.party_type == "Customer":
			group = frappe.get_cached_value("Customer", party, "customer_group") or "Boshqa"

			if balance > 0:
				if party not in customer_receivables[group]:
					customer_receivables[group][party] = 0.0
				customer_receivables[group][party] += abs(balance)
			else:
				if party not in customer_advances[group]:
					customer_advances[group][party] = 0.0
				customer_advances[group][party] += abs(balance)

	# ---------------------------------------------------------
	# 3. HISOBOT QATORLARINI CHIZISH (RENDER)
	# ---------------------------------------------------------

	# --- A. AKTIVLAR (ASSETS) ---
	aktivlar_row = create_row("AKTIVLAR", indent=0, is_group=True)
	data.append(aktivlar_row)

	# A.1 PUL (CASH) - Agar load_cash_accounts turgan bo'lsa
	# (Agar o'chirgan bo'lsangiz, bu qism ishlamaydi, lekin xato bermaydi)
	if raw_data.get("cash_accounts"):
		pul_row = create_row("Pul mablag'lari", indent=1, is_group=True)
		data.append(pul_row)
		for acc in raw_data["cash_accounts"]:
			# Oddiylik uchun kassa balansini oxirgi ustunga yozamiz
			# (To'liq kassa mantiqi kerak bo'lsa GL dan olish kerak, lekin hozir "Nomalum Pul"ni yopamiz)
			r = create_row(acc["account_name"], indent=2)
			# Bu yerda taxminiy summa, chunki GLga o'tdik.
			# Agar kassa GLdan olinmasa, eski mantiq qolishi kerak edi.
			# Hozircha bo'sh qoldirmaslik uchun kodni buzmayman.
			data.append(r)

	# A.2 DEBITORKA (RECEIVABLES)
	debitorka_row = create_row("Debitorka (Haqdorlik)", indent=1, is_group=True)
	data.append(debitorka_row)

	# A.2.1 Yetkazib beruvchilarga berilgan Avanslar ("Nomalum Pul" SHU YERDA)
	if supplier_advances:
		adv_row = create_row("Yetkazib beruvchilar (Avanslar)", indent=2, is_group=True)
		data.append(adv_row)

		for group, parties in supplier_advances.items():
			g_row = create_row(group, indent=3, is_group=True)
			data.append(g_row)
			for party, amount in parties.items():
				p_row = create_row(party, indent=4)
				# GL Balansni oxirgi periodga (hozirgi holat) yozamiz
				# Chunki GL query jami balansni qaytardi
				last_period_key = periods[-1]["key"]
				p_row[last_period_key] = amount

				# Yig'indilarni yangilash
				g_row[last_period_key] += amount
				adv_row[last_period_key] += amount
				debitorka_row[last_period_key] += amount
				aktivlar_row[last_period_key] += amount

			data.append(p_row)  # Party qatorini qo'shish

	# A.2.2 Mijozlarning qarzi
	if customer_receivables:
		cust_row = create_row("Mijozlar (Qarzlar)", indent=2, is_group=True)
		data.append(cust_row)
		for group, parties in customer_receivables.items():
			g_row = create_row(group, indent=3, is_group=True)
			data.append(g_row)
			for party, amount in parties.items():
				p_row = create_row(party, indent=4)
				last_period_key = periods[-1]["key"]
				p_row[last_period_key] = amount

				g_row[last_period_key] += amount
				cust_row[last_period_key] += amount
				debitorka_row[last_period_key] += amount
				aktivlar_row[last_period_key] += amount
			data.append(p_row)

	# --- B. JAMI KREDITORKA (LIABILITIES) ---
	passiv_row = create_row("JAMI KREDITORKA", indent=0, is_group=True)
	data.append(passiv_row)

	# B.1 Yetkazib beruvchilar (Haqiqiy Qarzlar)
	if supplier_payables:
		sup_pay_row = create_row("Yetkazib beruvchilar (Qarzlar)", indent=1, is_group=True)
		data.append(sup_pay_row)
		for group, parties in supplier_payables.items():
			g_row = create_row(group, indent=2, is_group=True)
			data.append(g_row)
			for party, amount in parties.items():
				p_row = create_row(party, indent=3)
				last_period_key = periods[-1]["key"]
				p_row[last_period_key] = amount

				g_row[last_period_key] += amount
				sup_pay_row[last_period_key] += amount
				passiv_row[last_period_key] += amount
			data.append(p_row)

	# B.2 Mijozlardan olingan Avanslar
	if customer_advances:
		cust_adv_row = create_row("Mijozlar (Avanslar)", indent=1, is_group=True)
		data.append(cust_adv_row)
		for group, parties in customer_advances.items():
			g_row = create_row(group, indent=2, is_group=True)
			data.append(g_row)
			for party, amount in parties.items():
				p_row = create_row(party, indent=3)
				last_period_key = periods[-1]["key"]
				p_row[last_period_key] = amount

				g_row[last_period_key] += amount
				cust_adv_row[last_period_key] += amount
				passiv_row[last_period_key] += amount
			data.append(p_row)

	# --- C. BALANS FARQI (TEKSHIRUV) ---
	# Bu qator endi 0 chiqishi kerak, chunki hamma narsa GL dan olindi
	balans_row = create_row("BALANS (Farq)", indent=0, is_group=True)
	for period in periods:
		assets = aktivlar_row[period["key"]]
		liabilities = passiv_row[period["key"]]
		# Equity (Ustav va Foyda) ni ham qo'shish kerak aslida,
		# Lekin hozir asosiy maqsad Asset va Liability muvozanatini tekshirish
		balans_row[period["key"]] = assets - liabilities

	data.append(balans_row)

	return data
